File: rec-sepval/src/main/java/net/kimleo/rec/sepval/parser/ParseState.java
package net.kimleo.rec.sepval.parser;

public class ParseState {
    private final String input;
    private int index = 0;

    public ParseState(String input) {
        this.input = input;
    }

    public int getSize() {
        return input.length();
    }

    public Character current() {
        return eof() ? null : input.charAt(index);
    }

    public boolean eof() {
        return index >= input.length();
    }

    public Character next() {
        index ++;
        return current();
    }

    public Character of(int index) {
        if (index >= getSize()) {
            return null;
        }
        return input.charAt(index);
    }
}

File: rec-sepval/src/main/kotlin/net/kimleo/rec/sepval/parser/ParseState.kt
package net.kimleo.rec.sepval.parser


class ParseState(val input: String) {
    val size = input.length
    var index = 0

    fun current(): Char? {
        return if (eof()) null else input[index]
    }

    fun eof(): Boolean {
        return index >= size
    }

    fun next(): Char? {
        index ++
        return if (eof()) null else current()
    }
}
